File: leetcode/minimum-number-of-arrows-to-burst-balloons.py
# 452. Minimum Number of Arrows to Burst Balloons
# 🟠 Medium
#
# https://leetcode.com/problems/minimum-number-of-arrows-to-burst-balloons/
#
# Tags: Array - Greedy - Sorting

import timeit
from typing import List


# Sort the balloons, then view them as intervals, each interval needs to
# be "hit" by, at least, one arrow. We can choose the first interval
# that has not been hit already and shot at its end point, then move to
# the next interval, if its start point is before/equal to the shot, it
# can be hit as well, we also need to check its end point and adjust
# the shot to make sure we also hit this new balloon shot = min(shot, end)
#
# Time complexity: O(n*log(n)) - The sorting step has the highest
# complexity, once sorted, we visit each balloon in the input once and
# do constant time operations with it.
# Space complexity: O(n) - Sorting in Python takes space.
#
# Runtime 1333 ms Beats 86.96%
# Memory 59.9 MB Beats 35.51%
class Solution:
    def findMinArrowShots(self, points: List[List[int]]) -> int:
        # Sort the intervals then greedily burst balloons.
        points.sort()
        # We are guaranteed to have, at least, one balloon, we need to
        # use one arrow to burst, greedily shot it at the end of the
        # interval trying to hit other balloons, we can move it back
        # later because the balloons are sorted.
        last_shot, arrows = points[0][1], 1
        # Iterate over the rest of the balloons finding the best shots
        # to burst them all.
        for start, end in points[1:]:
            # If this balloons start is before or the same as the last
            # shot, we can burst it with the same arrow.
            if start <= last_shot:
                # If the end of the balloon is before the last shot's
                # position, we need to move it back to make sure we hit
                # this balloon as well, we don't need to worry about the
                # start because balloons are sorted.
                if end < last_shot:
                    last_shot = end
            # If we could not hit this balloon using the previous arrow,
            # we will need to consume one more.
            else:
                last_shot = end
                arrows += 1
        return arrows


def test():
    executors = [Solution]
    tests = [
        [[[1, 5]], 1],
        [[[1, 5], [2, 5]], 1],
        [[[1, 2], [2, 5]], 1],
        [[[1, 2], [4, 5]], 2],
        [[[1, 2], [2, 3], [3, 4], [4, 5]], 2],
        [[[1, 2], [3, 4], [5, 6], [7, 8]], 4],
        [[[10, 16], [2, 8], [1, 6], [7, 12]], 2],
        [[[10, 16], [2, 8], [1, 6], [7, 12], [2, 4]], 2],
        [[[12, 16], [2, 7], [1, 6], [8, 11], [2, 4]], 3],
    ]
    for executor in executors:
        start = timeit.default_timer()
        for _ in range(1):
            for col, t in enumerate(tests):
                sol = executor()
                result = sol.findMinArrowShots(t[0])
                exp = t[1]
                assert result == exp, (
                    f"\033[93m» {result} <> {exp}\033[91m for"
                    + f" test {col} using \033[1m{executor.__name__}"
                )
        stop = timeit.default_timer()
        used = str(round(stop - start, 5))
        cols = "{0:20}{1:10}{2:10}"
        res = cols.format(executor.__name__, used, "seconds")
        print(f"\033[92m» {res}\033[0m")


test()

File: leetcode/minimum-number-of-arrows-to-burst-balloons.rs
// 452. Minimum Number of Arrows to Burst Balloons
// 🟠 Medium
//
// https://leetcode.com/problems/minimum-number-of-arrows-to-burst-balloons/
//
// Tags: Array - Greedy - Sorting

struct Interval {
    start: i32,
    end: i32,
}

struct Solution;
impl Solution {
    /// We sort the intervals by start, then iterate through them keeping track of the position at
    /// which we fired the last arrow until we cannot "reuse" it to burst the next ballon, then we
    /// increment the count of arrows used and start updating the position at which we used the
    /// arrow greedily using the furthest right value we can until we are forced to again use a new
    /// arrow.
    ///
    /// Time complexity: O(n*log(n)) - Sorting has the highest time complexity, after that O(n)
    /// Space complexity: O(1) - Constant extra space used if we ignore the mutable copy of the
    /// input array and the extra memory used by the sorting algorithm.
    ///
    /// Runtime 45 ms Beats 53%
    /// Memory 9.34 MB Beats 67%
    #[allow(dead_code)]
    pub fn find_min_arrow_shots_sort_by_start(points: Vec<Vec<i32>>) -> i32 {
        let mut res = 1;
        let mut points = points;
        points.sort_unstable();
        let mut last_arrow = points[0][1];
        for point in points.into_iter().skip(1) {
            if point[0] > last_arrow {
                last_arrow = point[1];
                res += 1;
            } else if point[1] < last_arrow {
                last_arrow = point[1];
            }
        }
        res
    }

    /// Similar solution but sort by end time and use an Interval struct to clean up the code.
    /// Sorting by end lets us not have to check if we need to modify the point at which we
    /// previously used the last arrow.
    ///
    /// Time complexity: O(n*log(n)) - Sorting has the highest time complexity, after that O(n)
    /// Space complexity: O(1) - Constant extra space used if we ignore the mutable copy of the
    /// input array and the extra memory used by the sorting algorithm.
    ///
    /// Runtime 37 ms Beats 90%
    /// Memory 9.48 MB Beats 45%
    pub fn find_min_arrow_shots(points: Vec<Vec<i32>>) -> i32 {
        let mut res = 1;
        let mut intervals = points
            .iter()
            .map(|v| Interval {
                start: v[0],
                end: v[1],
            })
            .collect::<Vec<_>>();
        intervals.sort_unstable_by_key(|i| i.end);
        let mut last_arrow = intervals[0].end;
        for interval in intervals.into_iter().skip(1) {
            if interval.start > last_arrow {
                last_arrow = interval.end;
                res += 1;
            }
        }
        res
    }
}

// Tests.
fn main() {
    let tests = [
        (vec![[1, 2], [3, 4], [5, 6], [7, 8]], 4),
        (vec![[1, 2], [2, 3], [3, 4], [4, 5]], 2),
        (vec![[10, 16], [2, 8], [1, 6], [7, 12]], 2),
        (
            vec![[9, 12], [1, 10], [4, 11], [8, 12], [3, 9], [6, 9], [6, 7]],
            2,
        ),
    ];
    println!("\n\x1b[92m» Running {} tests...\x1b[0m", tests.len());
    let mut success = 0;
    for (i, t) in tests.iter().enumerate() {
        let res = Solution::find_min_arrow_shots(t.0.iter().map(|a| a.to_vec()).collect());
        if res == t.1 {
            success += 1;
            println!("\x1b[92m✔\x1b[95m Test {} passed!\x1b[0m", i);
        } else {
            println!(
                "\x1b[31mx\x1b[95m Test {} failed expected: {:?} but got {}!!\x1b[0m",
                i, t.1, res
            );
        }
    }
    println!();
    if success == tests.len() {
        println!("\x1b[30;42m✔ All tests passed!\x1b[0m")
    } else if success == 0 {
        println!("\x1b[31mx \x1b[41;37mAll tests failed!\x1b[0m")
    } else {
        println!(
            "\x1b[31mx\x1b[95m {} tests failed!\x1b[0m",
            tests.len() - success
        )
    }
}
